core/src/main/java/com/radixdlt/logger/EventLoggerModule.java

```diff
@@ -175,11 +176,12 @@ EventProcessorOnDispatch<?> logRounds(Function<BFTValidatorId, String> nodeStrin
   @Singleton
   @SuppressWarnings("UnstableApiUsage")
   EventProcessorOnDispatch<?> ledgerUpdate(
-      @Self BFTValidatorId self, Function<ECDSASecp256k1PublicKey, String> nodeString) {
+      // The `Provider` indirection is needed here to break an unexpected circular dependency.
+      @Self Provider<BFTValidatorId> self, Function<ECDSASecp256k1PublicKey, String> nodeString) {
```

Comment:
Nice :). I don't even think this is that bad - because the EventLoggerModule feels like it should be near the bottom of the dependency tree, and the BFTValidatorId is near the top of the dependency tree (because it requires loading genesis) -- so if the `EventLoggerModule` wants to include `BFTValidatorId`, I think filling it in later feels reasonable 👍

Reply:
Yeah I guess "a logger" can be allowed to "peek" across layers (i.e. "inspect" different stuff that it conceptually should not depend on, only for logging/debugging purposes) 👍